### PR TITLE
Add Jaeger propagator

### LINKED ISSUE #542: Implement Jaeger Propagator
As per specs:
https://github.com/open-telemetry/opentelemetry-specification/blob/11cc73939a32e3a2e6f11bdeab843c61cf8594e9/specification/context/api-propagators.md#propagators-distribution

### PR DESCRIPTION
**Pull request by seemk**
Fixes #542 

## Changes

Adds Jaeger trace extraction and injection as specified [here](https://www.jaegertracing.io/docs/1.22/client-libraries/#propagation-format).

Baggage is not part of the PR as it's not supported by the otel-cpp API yet.

Miscellaneous:
* Removed `GetCurrentSpan` duplication from each propagator. Moved it to context's member function.
* Added hex (zero left-padded parse) and string helper (split) functions. These can be used in the b3 and http propagators to simplify code and remove duplication. Not part of this PR though, I will do a follow up if this one gets merged.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: api/include/opentelemetry/trace/propagation/jaeger.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    if (!span_context.IsValid())
+    {
+      return;
+    }
+
+    const size_t trace_id_length = 32;
+    const size_t span_id_length  = 16;
+
+    // trace-id(32):span-id(16):0:debug(2)
+    char trace_state[trace_id_length + span_id_length + 6];
```
[the remark this answers is not in the document]
Comment (seemk) [reply api/include/opentelemetry/trace/propagation/jaeger.h]: Renamed it to `trace_identity`

File: api/include/opentelemetry/context/context.h [hunk 2]
```diff
@@ -93,6 +93,17 @@ class Context
 
   bool operator==(const Context &other) const noexcept { return (head_ == other.head_); }
 
+  trace::SpanContext GetCurrentSpan() const
```
[the remark this answers is not in the document]
Comment (seemk) [reply api/include/opentelemetry/context/context.h]: Moved it to `propagation::detail`, but I think the context API is somewhat unfriendly to users, especially if accessing and setting spans is probably the most common use case.

File: api/include/opentelemetry/trace/propagation/jaeger.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  {
+    SpanContext span_context = ExtractImpl(getter, carrier);
+    nostd::shared_ptr<Span> sp{new DefaultSpan(span_context)};
+    return context.SetValue(kSpanKey, sp);
+  }
+
+private:
+  static TraceFlags GetTraceFlags(uint8_t jaeger_flags)
+  {
+    uint8_t sampled = jaeger_flags & 0x01;
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/jaeger.h:78]: nit: define the sampled bit as constant, like `kIsSampled` in the current class?

File: api/include/opentelemetry/trace/propagation/jaeger.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    nostd::string_view trace_fields[trace_field_count];
+
+    if (detail::SplitString(trace_identity, ':', trace_fields, trace_field_count) !=
+        trace_field_count)
+    {
+      return SpanContext::GetInvalid();
+    }
+
+    nostd::string_view trace_id_hex = trace_fields[0];
+    nostd::string_view span_id_hex  = trace_fields[1];
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/jaeger.h:96]: Is `trace_fields[2]` not used?

Comment (seemk) [reply api/include/opentelemetry/trace/propagation/jaeger.h:96]: Yup, it's the deprecated parent span ID (https://www.jaegertracing.io/docs/1.22/client-libraries/#value)

File: api/include/opentelemetry/trace/propagation/jaeger.h (hunk 3 above)
Comment (seemk) [reply api/include/opentelemetry/trace/propagation/jaeger.h]: Done!

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**